src/shared/local-addresses.c

```diff
@@ -313,3 +316,154 @@ int local_gateways(sd_netlink *context, int ifindex, int af, struct local_addres
+                 * really influencing the routing decision, i.e. this one works for us.*/
+                r = socket_bind_to_ifindex(fd, a[i].ifindex);
+                if (r < 0)
+                        log_debug_errno(r, "Failed to bind socket to interface %i, ignoring: %m", a[i].ifindex);
+
+                if (connect(fd, &sa.sa, SOCKADDR_LEN(sa)) < 0)
+                        log_debug_errno(errno, "Failed to connect SOCK_DGRAM socket to gateway, ignoring: %m");
+
+                /* Let's now read the socket address of the socket. A routing decision should have been
+                 * made. Let's verify that and use the data. */
+                salen = SOCKADDR_LEN(sa);
+                if (getsockname(fd, &sa.sa, &salen) < 0)
+                        return -errno;
+                assert(sa.sa.sa_family == a[i].family);
+                assert(salen == SOCKADDR_LEN(sa));
+
+                if (!b) {
+                        /* Allocate space for two addresses, since that's how much we need at max. Yes, this
+                         * might mean we allocate a bit too much in many cases, but it's not too bad. */
+                        b = new(struct local_address, 2);
+                        if (!b)
+                                return -ENOMEM;
+                }
+
+                assert(found <= 1);
+
+                switch (a[i].family) {
+
+                case AF_INET:
+                        if (in4_addr_is_null(&sa.in.sin_addr))
+                                continue;
+
+                        b[found++] = (struct local_address) {
+                                .family = a[i].family,
+                                .ifindex = a[i].ifindex,
+                                .address.in = sa.in.sin_addr,
+                        };
+
+                        break;
+
+                case AF_INET6:
+                        if (in6_addr_is_null(&sa.in6.sin6_addr))
+                                continue;
+
+                        b[found++] = (struct local_address) {
+                                .family = a[i].family,
+                                .ifindex = a[i].ifindex,
+                                .address.in6 = sa.in6.sin6_addr,
+                        };
+                        break;
+
+                default:
+                        assert_not_reached("Unexpected protocol");
+                }
+
+                if (found == 2) {
+                        /* We have now seen both protocols, we can leave */
+                        assert(done_af != AF_UNSPEC);
+                        assert(done_af != a[i].family);
+                        break;
+                }
+
+                /* We covered this address family now, let's remember that */
+                assert(found == 1);
+                assert(done_af == AF_UNSPEC);
+                done_af = a[i].family;
+        }
+
+        if (ret) {
```

Comment:
is there any point in doing the work if `ret` is not set? iow, shouldn't it be asserted?

Reply:
local_addresses() + locale_gateways() which work very similar support being called with NULL arguments too. We implement basically the same API, hence we should do so here too. And it might even make sense to call the function with ret=NULL, for example if all you want to check is if we have a routable local IP address or not. if we do the function will return > 0 after all, if we don't we'll return == 0. And that info alone is very useful I'd say.

(But of course, we don#t actually call it this way)

Reply:
ok, makes sense to me